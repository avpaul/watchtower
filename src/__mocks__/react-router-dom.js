import React from 'react';
import PropTypes from 'prop-types';

const rrd = require('react-router-dom');

// Just render plain div with its children
rrd.BrowserRouter = ({ children }) => <div>{children}</div>;

rrd.BrowserRouter.propTypes = {
  children: PropTypes.objectOf(PropTypes.element, PropTypes.func).isRequired,
};

module.exports = rrd;
