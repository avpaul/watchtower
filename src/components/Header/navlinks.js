import userIcon from '../../static/User.svg';
import userIconFill from '../../static/User-fill.svg';
import cogIcon from '../../static/Settings.svg';
import cogIconFill from '../../static/Settings-fill.svg';
import dashboardIcon from '../../static/Dashboard.svg';
import dashboardIconFill from '../../static/DashboardFill.svg';
import feedbackIcon from '../../static/Feedback.svg';
import feedbackIconFill from '../../static/FeedbackFill.svg';
import performanceIcon from '../../static/Perfomance.svg';
import performanceFillIcon from '../../static/PerfomanceFill.svg';
import historyIcon from '../../static/History.svg';
import historyFillIcon from '../../static/HistoryFill.svg';
import cadreProgramIcon from '../../static/cadreProgramIcon.svg';
import cadreProgramIconFill from '../../static/cadreProgramIconFill.svg';

export const menuOptions = {
  dashboard: {
    key: 'dashboard',
    name: 'Dashboard',
    path: '/dashboard',
    icon: dashboardIcon,
    activeIcon: dashboardIconFill
  },
  performance: {
    key: 'performance',
    name: 'My Performance',
    path: '/performance',
    icon: performanceIcon,
    activeIcon: performanceFillIcon
  },
  developers: {
    key: 'developers',
    name: 'Developers',
    path: '/developers',
    icon: userIcon,
    activeIcon: userIconFill
  },
  feedback: {
    key: 'feedback',
    name: 'Feedback',
    path: '/feedback',
    icon: feedbackIcon,
    activeIcon: feedbackIconFill
  },
  cadre: {
    key: 'cadre',
    name: 'Cadre Program',
    path: '/cadre/projects',
    set setPath(defaultPath) {
      this.path = `/cadre/${defaultPath}`;
    },
    get getPath() {
      return this.path;
    },
    icon: cadreProgramIcon,
    activeIcon: cadreProgramIconFill
  },
  history: {
    key: 'history',
    name: 'My History',
    icon: historyIcon,
    activeIcon: historyFillIcon
  }
};

export const items = [
  {
    key: 'fellows',
    name: 'Fellows',
    path: '#',
    icon: userIcon,
    activeIcon: userIconFill,
    dropdown: 'dropdown'
  },
  {
    key: 'dashboard',
    name: 'Dashboard',
    path: '#',
    icon: cogIcon,
    activeIcon: cogIconFill
  }
];

export const fellowItems = [
  menuOptions.dashboard,
  menuOptions.performance,
  menuOptions.history
];

export const managerItems = [
  menuOptions.dashboard,
  menuOptions.developers,
  menuOptions.feedback
];

export const managerWithCadreItems = [
  menuOptions.dashboard,
  menuOptions.developers,
  menuOptions.feedback,
  menuOptions.cadre
];

export const cadreManagerItems = [menuOptions.cadre];

export const opsItems = [
  menuOptions.dashboard,
  menuOptions.developers,
  menuOptions.feedback,
  menuOptions.cadre
];

export const getMenuItems = (role, roles = null) => {
  switch (role) {
    case 'Fellow':
      return fellowItems;
    case 'WATCH_TOWER_TTL':
    case 'WATCH_TOWER_LF':
    case 'WATCH_TOWER_EM':
    case 'WATCH_TOWER_SL':
      return roles && roles.includes('CADRE_TEAM_MANAGER')
        ? managerWithCadreItems
        : managerItems;
    case 'WATCH_TOWER_OPS':
      return opsItems;
    case 'CADRE_TEAM_MANAGER':
      return cadreManagerItems;
    default:
      return fellowItems;
  }
};
