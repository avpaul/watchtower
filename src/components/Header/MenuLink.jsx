import React from 'react';
import { Link } from "react-router-dom";
import PropTypes from 'prop-types';

/**
 *
 * MenuLink Component
 * @param {Object} props
 *
 * @returns {JSX} React JSX
 */
const MenuLink = ({
  link, handleMenuClick, isActive, role,
}) => {
  let active = '';
  // if the link state is active, change icon
  let iconImg = link.icon;
  if (isActive) {
    active = 'active';
    iconImg = link.activeIcon || link.icon;
  }

  function renderHeader() {
    const path = link.path ? link.path: ''; 
    return (
      <Link
        className={`navlink ${active}`}
        data-link-key={link.key}
        onClick={handleMenuClick}
        to={path}
        role="link"
        onKeyPress={handleMenuClick}
        tabIndex="0"
      >
       <span className="navicon">
          <img className={`${link.key}__icon`} src={iconImg} alt={`${link.key}Icon`} />
        </span>
        {link.name}
      </Link>
    );
  }

  if (role === 'Fellow') {
    return (
      <div className="menulink">
        {(renderHeader())}
      </div>
    );
  }


  if (role === 'WATCH_TOWER_OPS' || 'WATCH_TOWER_EM' || 'WATCH_TOWER_TTL' || 'WATCH_TOWER_LF') {
    return (
      <div className="menulink">
        {(renderHeader())}
        {link.dropdown !== undefined
          && (
          <div>
            <a className="nav-link dropdown-toggle" href="/" data-toggle="dropdown">
              <div className="dropdown-menu dropdown-menu-right">
                {link.dropdown.map(
                  item => (
                    <a className="dropdown-item" href="/">{item}</a>
                  ),
                )}
              </div>
            </a>
          </div>
          )
          }
      </div>
    );
  }
  return null;
};

MenuLink.propTypes = {
  link: PropTypes.shape({
    key: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    path: PropTypes.string.isRequired,
    icon: PropTypes.node.isRequired,
    activeIcon: PropTypes.node,
  }).isRequired,
  handleMenuClick: PropTypes.func.isRequired,
  isActive: PropTypes.bool.isRequired,
  role: PropTypes.string.isRequired,
};

export default MenuLink;
