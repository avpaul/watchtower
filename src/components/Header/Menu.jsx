import React from 'react';
import PropTypes from 'prop-types';
import MenuLink from './MenuLink';

/**
 *
 * Menu Component
 * @param {Object} props
 *
 * @returns {JSX} React JSX
 */
const Menu = ({ items, handleMenuClick, activeItems }) => (
  <div className="header--bottom d-flex justify-content-center">
    {items.map(
      item => (
        <MenuLink
          key={item.key}
          link={item}
          handleMenuClick={handleMenuClick}
          isActive={activeItems[item.key]}
        />
      ),
    )
    }
  </div>
);

Menu.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleMenuClick: PropTypes.func.isRequired,
  activeItems: PropTypes.shape({
    fellows: PropTypes.bool.isRequired,
    settings: PropTypes.bool.isRequired,
  }).isRequired,
};

export default Menu;
