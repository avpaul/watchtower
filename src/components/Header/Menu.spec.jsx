import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { getMenuItems } from './navlinks';
import Menu from './Menu';
import MenuLink from '.';

const links = [
  {
    key: 'fellows',
    name: 'Fellows',
    path: '#',
    icon: <div className="default-icon" />,
    activeIcon: <div className="active-icon" />,
    dropdown: ['All Developers', 'Developers under PIP']
  },
  {
    key: 'dashboard',
    name: 'Dashboard',
    path: '#',
    icon: <div className="default-icon" />,
    activeIcon: <div className="active-icon" />,
    dropdown: []
  }
];

const props = {
  items: links,
  handleMenuClick: jest.fn(),
  activeItems: { fellows: false, dashboard: true, settings: true },
  role: 'WATCH_TOWER_LF'
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <MemoryRouter>
      <Menu {...props} />
    </MemoryRouter>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

it('renders MenuLink child components', () => {
  const wrapper = shallow(<Menu {...props} />);
  expect(wrapper.find(MenuLink)).toBeDefined();
});

describe('Menu Component Test Suite', () => {
  it('renders witout error', () => {
    const items = getMenuItems('Fellow');
    const activeItems = [
      {
        key: 'dashboard',
        name: 'Dashboard',
        path: '/dashboard'
      }
    ];

    const snap = shallow(
      <Menu items={items} role="Fellow" activeItems={activeItems} />
    );
    expect(snap).toMatchSnapshot();
  });
});
