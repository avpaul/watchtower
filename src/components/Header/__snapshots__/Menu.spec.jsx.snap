// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Component Test Suite renders witout error 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Menu
    activeItems={
        Array [
            Object {
              "key": "dashboard",
              "name": "Dashboard",
              "path": "/dashboard",
            },
          ]
    }
    items={
        Array [
            Object {
              "activeIcon": "DashboardFill.svg",
              "icon": "Dashboard.svg",
              "key": "dashboard",
              "name": "Dashboard",
            },
            Object {
              "activeIcon": "PerfomanceFill.svg",
              "icon": "Perfomance.svg",
              "key": "Performance",
              "name": "My Performance",
            },
            Object {
              "activeIcon": "HistoryFill.svg",
              "icon": "History.svg",
              "key": "History",
              "name": "My History",
            },
          ]
    }
    role="Fellow"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <MenuLink
          handleMenuClick={undefined}
          isActive={undefined}
          link={
                    Object {
                              "activeIcon": "DashboardFill.svg",
                              "icon": "Dashboard.svg",
                              "key": "dashboard",
                              "name": "Dashboard",
                            }
          }
          role="Fellow"
/>,
        <MenuLink
          handleMenuClick={undefined}
          isActive={undefined}
          link={
                    Object {
                              "activeIcon": "PerfomanceFill.svg",
                              "icon": "Perfomance.svg",
                              "key": "Performance",
                              "name": "My Performance",
                            }
          }
          role="Fellow"
/>,
        <MenuLink
          handleMenuClick={undefined}
          isActive={undefined}
          link={
                    Object {
                              "activeIcon": "HistoryFill.svg",
                              "icon": "History.svg",
                              "key": "History",
                              "name": "My History",
                            }
          }
          role="Fellow"
/>,
      ],
      "className": "header--bottom d-flex justify-content-center",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "dashboard",
        "nodeType": "function",
        "props": Object {
          "handleMenuClick": undefined,
          "isActive": undefined,
          "link": Object {
            "activeIcon": "DashboardFill.svg",
            "icon": "Dashboard.svg",
            "key": "dashboard",
            "name": "Dashboard",
          },
          "role": "Fellow",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Performance",
        "nodeType": "function",
        "props": Object {
          "handleMenuClick": undefined,
          "isActive": undefined,
          "link": Object {
            "activeIcon": "PerfomanceFill.svg",
            "icon": "Perfomance.svg",
            "key": "Performance",
            "name": "My Performance",
          },
          "role": "Fellow",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "History",
        "nodeType": "function",
        "props": Object {
          "handleMenuClick": undefined,
          "isActive": undefined,
          "link": Object {
            "activeIcon": "HistoryFill.svg",
            "icon": "History.svg",
            "key": "History",
            "name": "My History",
          },
          "role": "Fellow",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <MenuLink
            handleMenuClick={undefined}
            isActive={undefined}
            link={
                        Object {
                                    "activeIcon": "DashboardFill.svg",
                                    "icon": "Dashboard.svg",
                                    "key": "dashboard",
                                    "name": "Dashboard",
                                  }
            }
            role="Fellow"
/>,
          <MenuLink
            handleMenuClick={undefined}
            isActive={undefined}
            link={
                        Object {
                                    "activeIcon": "PerfomanceFill.svg",
                                    "icon": "Perfomance.svg",
                                    "key": "Performance",
                                    "name": "My Performance",
                                  }
            }
            role="Fellow"
/>,
          <MenuLink
            handleMenuClick={undefined}
            isActive={undefined}
            link={
                        Object {
                                    "activeIcon": "HistoryFill.svg",
                                    "icon": "History.svg",
                                    "key": "History",
                                    "name": "My History",
                                  }
            }
            role="Fellow"
/>,
        ],
        "className": "header--bottom d-flex justify-content-center",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "dashboard",
          "nodeType": "function",
          "props": Object {
            "handleMenuClick": undefined,
            "isActive": undefined,
            "link": Object {
              "activeIcon": "DashboardFill.svg",
              "icon": "Dashboard.svg",
              "key": "dashboard",
              "name": "Dashboard",
            },
            "role": "Fellow",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Performance",
          "nodeType": "function",
          "props": Object {
            "handleMenuClick": undefined,
            "isActive": undefined,
            "link": Object {
              "activeIcon": "PerfomanceFill.svg",
              "icon": "Perfomance.svg",
              "key": "Performance",
              "name": "My Performance",
            },
            "role": "Fellow",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "History",
          "nodeType": "function",
          "props": Object {
            "handleMenuClick": undefined,
            "isActive": undefined,
            "link": Object {
              "activeIcon": "HistoryFill.svg",
              "icon": "History.svg",
              "key": "History",
              "name": "My History",
            },
            "role": "Fellow",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
